raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java

```diff
@@ -0,0 +1,489 @@
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        Map<ListenerName, InetSocketAddress> anotherEndpointMap = new HashMap<>();
+        anotherEndpointMap.put(
+                ListenerName.normalised("another"),
+                InetSocketAddress.createUnresolved("localhost", 9093)
+        );
+        Endpoints differentEndpoints = Endpoints.fromInetSocketAddresses(anotherEndpointMap);
+
+        assertNotEquals(endpoints, differentEndpoints);
+        assertNotEquals(endpoints.hashCode(), differentEndpoints.hashCode());
+    }
+
+    @Test
+    void testToString() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        ListenerName listener = ListenerName.normalised("listener");
+        ListenerName listener1 = ListenerName.normalised("listener1");
+        InetSocketAddress address = InetSocketAddress.createUnresolved("localhost", 9092);
+        InetSocketAddress address1 = InetSocketAddress.createUnresolved("localhost", 9093);
+        endpointMap.put(listener, address);
+        endpointMap.put(listener1, address1);
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        String expectedString = String.format(
+                "Endpoints(endpoints={%s=%s, %s=%s})",
+                listener,
+                address,
+                listener1,
+                address1
+        );
+
+        assertEquals(expectedString, endpoints.toString());
+    }
+
+    @Test
+    void testToBeginQuorumEpochRequestWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        BeginQuorumEpochRequestData.LeaderEndpointCollection leaderEndpoints = endpoints.toBeginQuorumEpochRequest();
+
+        assertEquals(1, leaderEndpoints.size());
+
+        Iterator<BeginQuorumEpochRequestData.LeaderEndpoint> iterator = leaderEndpoints.iterator();
+        BeginQuorumEpochRequestData.LeaderEndpoint endpoint = iterator.next();
+        assertEquals(ListenerName.normalised("listener").value(), endpoint.name());
+        assertEquals("localhost", endpoint.host());
+        assertEquals(9092, endpoint.port());
+    }
+
+    @Test
+    void testToBeginQuorumEpochRequestWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        BeginQuorumEpochRequestData.LeaderEndpointCollection leaderEndpoints = endpoints.toBeginQuorumEpochRequest();
+
+        assertEquals(0, leaderEndpoints.size());
```

Comment:
`leaderEndpoints` in the assert can be replaced by its value `endpoints.toBeginQuorumEpochRequest()`

Reply:
updated